{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "41011c57-da78-4675-a61a-74fa2a878a4b",
   "metadata": {},
   "source": [
    "# Spatial visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "673d444b-63a5-45d7-8f4c-d2b0da5822ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Packages\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import warnings \n",
    "warnings.filterwarnings('ignore')\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "from skimage.filters import threshold_otsu, gaussian\n",
    "from skimage.morphology import remove_small_objects\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "from anndata import AnnData, concat\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "# Customized packages\n",
    "from starmap.utilities import *\n",
    "from starmap.sequencing import *\n",
    "# from starmap.obj import STARMapDataset, load_data\n",
    "# import starmap.analyze as anz\n",
    "# import starmap.viz as viz\n",
    "\n",
    "import starmap.sc_util as su\n",
    "\n",
    "# test()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53f43d25-ea77-4230-b1c0-219ba2de9650",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get functions \n",
    "\n",
    "import colorsys\n",
    "from random import shuffle\n",
    "\n",
    "def intervals(parts, start_point, end_point):\n",
    "    duration = end_point - start_point\n",
    "    part_duration = duration / parts\n",
    "    return [((i * part_duration + (i + 1) * part_duration)/2) + start_point for i in range(parts)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca465718-2994-4eae-9110-32f239f94bee",
   "metadata": {
    "tags": []
   },
   "source": [
    "## IO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8e2a56b-ea17-4c69-9db7-1241f5ae7632",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set path\n",
    "base_path = 'Z:/Data/Analyzed/2021-11-23-Hu-MouseBrain/'\n",
    "out_path = os.path.join(base_path, 'output')\n",
    "fig_path = os.path.join(base_path, 'figures')\n",
    "\n",
    "out_path = os.path.join(base_path, 'output')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)\n",
    "    \n",
    "fig_path = os.path.join(base_path, 'figures')\n",
    "if not os.path.exists(fig_path):\n",
    "    os.mkdir(fig_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "598fbeef-b410-4309-9c34-530efd12c4ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "date = datetime.today().strftime('%Y-%m-%d')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6edfd3d-f75d-497a-9061-35c5d1a9f77d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.figdir = fig_path\n",
    "sc.set_figure_params(format='tif', dpi=150)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffd88876-d0ff-4563-82d2-3e1c47b34bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2eef63e-3bd4-41bc-9a2f-8ac95142a0dc",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b06242-1536-4a0d-a657-0ae04089685e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load new data\n",
    "adata = sc.read_h5ad(os.path.join(out_path, '2022-04-24-Hu-TissueRIBOmap-level3.h5ad'))\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f0d32f4-875d-4694-80d1-6640ef7daf90",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['level_1_colors'] = np.array(['#eded58', '#356be8'], dtype=object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054b8155-b0e8-4d4f-a5d0-693300b1f695",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['level_1_color_list'] = np.array(['#eded58', '#356be8'], dtype=object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb1df4b-0f2f-47ac-bffd-b47414162527",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "date = datetime.today().strftime('%Y-%m-%d')\n",
    "adata.write_h5ad(f\"{out_path}/2022-04-24-Hu-TissueRIBOmap-level3.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9127c349-e1e1-484a-9ef9-ded736b974a1",
   "metadata": {},
   "source": [
    "orig resolution: 0.09 micron / pixel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2fd3d4b-1bff-47c5-8ca6-7adfa3d284ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = round(23215 / 0.3 * 0.09 / 1000, 5)\n",
    "y = round(18332 / 0.3 * 0.09 / 1000, 5)\n",
    "print(f'X: {x} mm - Y: {y} mm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e18b90f5-a542-48c5-ad6a-0fb292e12056",
   "metadata": {},
   "source": [
    "## Colors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4b7223e-0268-4b30-b08f-ab4a1038b50d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### level-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c750dc6b-3ead-4ee0-bb14-191a1b0a908e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get colormap\n",
    "level_3_colors = adata.uns['level_3_color_list']\n",
    "level_3_pl = sns.color_palette(level_3_colors)\n",
    "level_3_cmap = ListedColormap(level_3_pl.as_hex())\n",
    "\n",
    "level_3_order = adata.uns['level_3_order']\n",
    "sns.palplot(level_3_pl)\n",
    "plt.xticks(range(len(level_3_order)), level_3_order, size=10, rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7601f589-5eb0-44b3-aca3-6acefb2a00c6",
   "metadata": {},
   "source": [
    "## Spatial map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84261d7f-e0c7-4078-ba5d-29325760a8d2",
   "metadata": {},
   "source": [
    "### scatter plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d2a81a1-a1e2-4191-882b-08255b13ab6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_as = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84eef129-06e7-4baf-a4c0-ff240851e4a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustermap_center_path = os.path.join(base_path, 'RIBOmap', 'cell_center_polished.csv')\n",
    "cell_center_df = pd.read_csv(clustermap_center_path, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e33efc9-c139-4ba7-a7ed-b9b9cf09e4a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.set_figure_params(format='tif', dpi=150)\n",
    "\n",
    "dot_size = 12\n",
    "fig, ax = plt.subplots(figsize=(18, 23))\n",
    "sns.scatterplot(x='column', y='row', data=cell_center_df, color='#787878', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "sns.scatterplot(x='column', y='row', hue='level_3', data=adata.obs, palette=level_3_pl, s=dot_size, edgecolor=None, legend=False, ax=ax)\n",
    "ax.axis('off')\n",
    "plt.tight_layout(pad=0)\n",
    "\n",
    "if save_as:\n",
    "    plt.savefig(os.path.join(fig_path, 'spatial-map-level-3.tif'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7516ba4f-9e16-4bb5-a8bf-d4a3feaf4052",
   "metadata": {},
   "source": [
    "### label image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "078c732b-23f4-4292-bb01-4d275f9cd5de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create label image from clustermap output\n",
    "from skimage.segmentation import expand_labels\n",
    "\n",
    "# load clustermap results\n",
    "reads_file_path = os.path.join(base_path, 'RIBOmap', 'remain_reads_polished.csv')\n",
    "reads_df = pd.read_csv(reads_file_path, index_col=0)\n",
    "\n",
    "center_file_path = os.path.join(base_path, 'RIBOmap', 'cell_center_polished.csv')\n",
    "centers_df = pd.read_csv(center_file_path, index_col=0)\n",
    "\n",
    "# modify reads filexh\n",
    "reads_df['spot_location_1'] = reads_df['spot_location_1'] * 0.3\n",
    "reads_df['spot_location_2'] = reads_df['spot_location_2'] * 0.3\n",
    "reads_df['spot_location_1'] = reads_df['spot_location_1'].astype(int)\n",
    "reads_df['spot_location_2'] = reads_df['spot_location_2'].astype(int)\n",
    "\n",
    "row = reads_df['spot_location_2'].max() + 1\n",
    "col = reads_df['spot_location_1'].max() + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbd7ad55-5d5b-4b81-9db2-857be400c30b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_barcode_dict = dict(zip(centers_df['cell_barcode'].values, centers_df.index.values))\n",
    "reads_df['orig_index'] = reads_df['cell_barcode'].map(cell_barcode_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44d0c32a-aac1-469c-a3eb-d84f49b85624",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create label image \n",
    "label_img = np.zeros([row, col], dtype=np.uint32)\n",
    "label_img[reads_df['spot_location_2'].values, reads_df['spot_location_1'].values] = reads_df['orig_index'].values + 1\n",
    "# label_img = expand_labels(label_img, distance=1)\n",
    "\n",
    "label_img_out_path = os.path.join(base_path, 'RIBOmap', 'labeled_cells_no_extend.tif')\n",
    "tifffile.imwrite(label_img_out_path, label_img)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f824c264-74aa-4ee9-a584-cde86118cf2d",
   "metadata": {},
   "source": [
    "### polygon map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ccbc032-87ae-4587-af46-2756b1749ae0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load segmentation \n",
    "current_seg_path = os.path.join(base_path, 'RIBOmap', 'labeled_cells_no_extend.tif')\n",
    "current_img = tifffile.imread(current_seg_path)\n",
    "\n",
    "# Store the images to adata object\n",
    "current_key = f\"RIBOmap_morph\"\n",
    "adata.uns[current_key] = {}\n",
    "adata.uns[current_key]['label_img'] = current_img"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a51b928-807a-4822-89a9-7b2e2f43c832",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Contruct polygon\n",
    "adata.uns[current_key]['qhulls'], adata.uns[current_key]['coords'], adata.uns[current_key]['centroids'] = su.get_qhulls_test(adata.uns[current_key]['label_img'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9e5c0cf-2658-4bad-9ffc-b97f440ac609",
   "metadata": {},
   "source": [
    "#### level-3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6ec6106-e8d4-490e-ac39-18e27b6040db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add data\n",
    "adata.uns[current_key]['good_cells'] = adata.obs.index.astype(int).values\n",
    "adata.uns[current_key]['colors'] = adata.obs.loc[:, 'level_3'].cat.codes.values\n",
    "\n",
    "# Plot\n",
    "su.plot_poly_cells_cluster_by_sample(adata, 'RIBOmap', level_3_cmap, show_plaque=False, show_tau=False, linewidth=0.5,\n",
    "                                        figscale=3, width=10, height=10, bg_color='#787878', save_as_real_size=True,\n",
    "                                         save=True, show=False, output_dir=fig_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aee06544-0813-4714-8a7c-be8748984b8e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Glia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f82a46-dd67-4251-a160-473385b7bd07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset adata\n",
    "sdata = adata[adata.obs['level_1'] == 'Glia', ]\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b07e541-36f5-405a-ba9b-847fd4f2da9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get colormap\n",
    "glia_order = []\n",
    "glia_colors = []\n",
    "\n",
    "for i, current_type in enumerate(sdata.uns['level_3_order']):\n",
    "    if current_type in sdata.obs['level_3'].unique():\n",
    "        glia_order.append(current_type)\n",
    "        glia_colors.append(sdata.uns['level_3_color_list'][i])\n",
    "        \n",
    "        \n",
    "glia_pl = sns.color_palette(glia_colors)\n",
    "glia_cmap = ListedColormap(glia_pl.as_hex())\n",
    "sns.palplot(glia_pl)\n",
    "plt.xticks(range(len(glia_order)), glia_order, size=10, rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68d3021d-65bb-4410-9c39-a25f1c585c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_fig_path = os.path.join(fig_path, 'Glia', 'sct')\n",
    "if not os.path.exists(current_fig_path):\n",
    "    os.mkdir(current_fig_path)\n",
    "\n",
    "# Add data\n",
    "sdata.uns[current_key]['good_cells'] = sdata.obs.index.astype(int).values\n",
    "sdata.uns[current_key]['colors'] = sdata.obs.loc[:, 'level_3'].cat.codes.values\n",
    "\n",
    "# Plot\n",
    "su.plot_poly_cells_cluster_by_sample(sdata, 'RIBOmap', glia_cmap, show_plaque=False, show_tau=False, linewidth=0.5,\n",
    "                                        figscale=3, width=10, height=10, save_as_real_size=True,\n",
    "                                         save=True, show=False, output_dir=current_fig_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f7c37d8-2d03-4fa9-9cd9-168a890fc107",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Excitatory neuron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff680501-a571-4455-80fb-0d9e1c5cddc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset adata\n",
    "sdata = adata[adata.obs['level_2'] == 'Excitatory neuron', ]\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcf675c0-4021-4cde-b9d5-2fdea8748893",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get colormap\n",
    "ex_order = []\n",
    "ex_colors = []\n",
    "\n",
    "for i, current_type in enumerate(sdata.uns['level_3_order']):\n",
    "    if current_type in sdata.obs['level_3'].unique():\n",
    "        ex_order.append(current_type)\n",
    "        ex_colors.append(sdata.uns['level_3_color_list'][i])\n",
    "        \n",
    "        \n",
    "ex_pl = sns.color_palette(ex_colors)\n",
    "ex_cmap = ListedColormap(ex_pl.as_hex())\n",
    "sns.palplot(ex_pl)\n",
    "plt.xticks(range(len(ex_order)), ex_order, size=10, rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c141d73-836f-4fe5-a6d8-7a4f0b5c80dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_fig_path = os.path.join(fig_path, 'Excitatory neuron', 'sct')\n",
    "if not os.path.exists(current_fig_path):\n",
    "    os.mkdir(current_fig_path)\n",
    "\n",
    "# Add data\n",
    "sdata.uns[current_key]['good_cells'] = sdata.obs.index.astype(int).values\n",
    "sdata.uns[current_key]['colors'] = sdata.obs.loc[:, 'level_3'].cat.codes.values\n",
    "\n",
    "# Plot\n",
    "su.plot_poly_cells_cluster_by_sample(sdata, 'RIBOmap', ex_cmap, show_plaque=False, show_tau=False, linewidth=0.5,\n",
    "                                        figscale=3, width=10, height=10, save_as_real_size=True,\n",
    "                                         save=True, show=False, output_dir=current_fig_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef1afad-fafe-4f4f-832b-8fbd68415520",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Inhibitory neuron"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "367906b4-3d49-4f8f-adef-c4cb489ff341",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset adata\n",
    "sdata = adata[adata.obs['level_2'] == 'Inhibitory neuron', ]\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d42bbe-b49d-4ba7-a5a9-c0c67da30975",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get colormap\n",
    "inh_order = []\n",
    "inh_colors = []\n",
    "\n",
    "for i, current_type in enumerate(sdata.uns['level_3_order']):\n",
    "    if current_type in sdata.obs['level_3'].unique():\n",
    "        inh_order.append(current_type)\n",
    "        inh_colors.append(sdata.uns['level_3_color_list'][i])\n",
    "        \n",
    "        \n",
    "inh_pl = sns.color_palette(inh_colors)\n",
    "inh_cmap = ListedColormap(inh_pl.as_hex())\n",
    "sns.palplot(inh_pl)\n",
    "plt.xticks(range(len(inh_order)), inh_order, size=10, rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96d7be9e-ac44-489e-a8a5-56f67592f7d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_fig_path = os.path.join(fig_path, 'Inhibitory neuron', 'sct')\n",
    "if not os.path.exists(current_fig_path):\n",
    "    os.mkdir(current_fig_path)\n",
    "\n",
    "# Add data\n",
    "sdata.uns[current_key]['good_cells'] = sdata.obs.index.astype(int).values\n",
    "sdata.uns[current_key]['colors'] = sdata.obs.loc[:, 'level_3'].cat.codes.values\n",
    "\n",
    "# Plot\n",
    "su.plot_poly_cells_cluster_by_sample(sdata, 'RIBOmap', inh_cmap, show_plaque=False, show_tau=False, linewidth=0.5,\n",
    "                                        figscale=3, width=10, height=10, save_as_real_size=True,\n",
    "                                         save=True, show=False, output_dir=current_fig_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf559abb-c810-4a98-b94d-1d8af03cef81",
   "metadata": {},
   "source": [
    "### polygon map with dots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eed8c844-532d-4508-8957-b942249270ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load reads \n",
    "good_reads = pd.read_csv(os.path.join(base_path, 'RIBOmap', 'remain_reads_polished.csv'), index_col=0)\n",
    "background_reads = pd.read_csv(os.path.join(base_path, 'RIBOmap', 'background_reads_polished.csv'), index_col=0)\n",
    "cell_center_df = pd.read_csv(os.path.join(base_path, 'RIBOmap', 'cell_center_polished.csv'), index_col=0)\n",
    "\n",
    "all_reads = pd.concat([good_reads, background_reads])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d25fd85-127c-4c1c-8223-92e9cd6e7d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# modify reads filexh\n",
    "good_reads['spot_location_1'] = good_reads['spot_location_1'] * 0.3\n",
    "good_reads['spot_location_2'] = good_reads['spot_location_2'] * 0.3\n",
    "good_reads['spot_location_1'] = good_reads['spot_location_1'].astype(int)\n",
    "good_reads['spot_location_2'] = good_reads['spot_location_2'].astype(int)\n",
    "\n",
    "row = good_reads['spot_location_2'].max() + 1\n",
    "col = good_reads['spot_location_1'].max() + 1\n",
    "\n",
    "background_reads['spot_location_1'] = background_reads['spot_location_1'] * 0.3\n",
    "background_reads['spot_location_2'] = background_reads['spot_location_2'] * 0.3\n",
    "background_reads['spot_location_1'] = background_reads['spot_location_1'].astype(int)\n",
    "background_reads['spot_location_2'] = background_reads['spot_location_2'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bffb740-1a69-4a8f-af80-0d3b36016b06",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_barcode_dict = dict(zip(cell_center_df['cell_barcode'].values, cell_center_df.index.values))\n",
    "good_reads['orig_index'] = good_reads['cell_barcode'].map(cell_barcode_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3642d054-166f-4418-8338-87fddb5308be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create scaled coords\n",
    "adata.obs['column_scaled'] = adata.obs['column'] * .3\n",
    "adata.obs['row_scaled'] = adata.obs['row'] * .3\n",
    "adata.obs['column_scaled'] = adata.obs['column_scaled'].astype(int)\n",
    "adata.obs['row_scaled'] = adata.obs['row_scaled'].astype(int)\n",
    "\n",
    "cell_center_df['column_scaled'] = cell_center_df['column'] * .3\n",
    "cell_center_df['row_scaled'] = cell_center_df['row'] * .3\n",
    "cell_center_df['column_scaled'] = cell_center_df['column_scaled'].astype(int)\n",
    "cell_center_df['row_scaled'] = cell_center_df['row_scaled'].astype(int)\n",
    "\n",
    "good_reads['column_scaled'] = good_reads['spot_location_1']\n",
    "good_reads['row_scaled'] = good_reads['spot_location_2']\n",
    "\n",
    "background_reads['column_scaled'] = background_reads['spot_location_1']\n",
    "background_reads['row_scaled'] = background_reads['spot_location_2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "542c0775-c023-4138-96f9-84f5aedb3552",
   "metadata": {},
   "outputs": [],
   "source": [
    "box_length = 5000\n",
    "# row_start = 18000\n",
    "# col_start = 7500\n",
    "\n",
    "row_start = 15000\n",
    "col_start = 8000\n",
    "\n",
    "row_end = row_start + box_length\n",
    "col_end = col_start + box_length\n",
    "\n",
    "dot_size = 7\n",
    "fig, ax = plt.subplots(figsize=(18, 23))\n",
    "sns.scatterplot(x='column_scaled', y='row_scaled', data=cell_center_df, color='#787878', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "\n",
    "rect = patches.Rectangle((col_start, row_start), box_length, box_length, linewidth=1, edgecolor='r', facecolor='none')\n",
    "ax.add_patch(rect)\n",
    "# plt.axis('off')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e64a96-b074-44a9-b7ff-a5ba021230f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create subsets \n",
    "adata_logical = (adata.obs['column_scaled'].isin(range(col_start, col_end))) & (adata.obs['row_scaled'].isin(range(row_start, row_end)))\n",
    "sdata = adata[adata_logical, :]\n",
    "\n",
    "reads_logical = (good_reads['column_scaled'].isin(range(col_start, col_end))) & (good_reads['row_scaled'].isin(range(row_start, row_end)))\n",
    "sreads_df = good_reads.loc[reads_logical, :]\n",
    "\n",
    "bg_reads_logical = (background_reads['column_scaled'].isin(range(col_start, col_end))) & (background_reads['row_scaled'].isin(range(row_start, row_end)))\n",
    "bgreads_df = background_reads.loc[bg_reads_logical, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0e9b2ba-b909-450a-a225-648f72f1422a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check subset\n",
    "dot_size = 7\n",
    "fig, ax = plt.subplots(figsize=(18, 23))\n",
    "\n",
    "sns.scatterplot(x='column_scaled', y='row_scaled', data=cell_center_df, color='#787878', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "sns.scatterplot(x='column_scaled', y='row_scaled', data=sdata.obs, color='red', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "sns.scatterplot(x='column_scaled', y='row_scaled', data=sreads_df, color='blue', s=.1, legend=False, edgecolor=None, ax=ax)\n",
    "sns.scatterplot(x='column_scaled', y='row_scaled', data=bgreads_df, color='green', s=.1, legend=False, edgecolor=None, ax=ax)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09436e04-70f5-4915-9d88-e5668af1150b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset image \n",
    "sdata.uns['RIBOmap_morph']['label_img_crop'] = sdata.uns['RIBOmap_morph']['label_img'][row_start:row_end, col_start:col_end]\n",
    "sdata.uns['RIBOmap_morph']['qhulls'], sdata.uns['RIBOmap_morph']['coords'], sdata.uns['RIBOmap_morph']['centroids'] = su.get_qhulls_test(sdata.uns['RIBOmap_morph']['label_img_crop'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "377ef23b-cde0-46a8-8222-b688c06e1cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata.obs['column_crop'] = sdata.obs['column_scaled'] - sdata.obs['column_scaled'].min()\n",
    "sdata.obs['row_crop'] = sdata.obs['row_scaled'] - sdata.obs['row_scaled'].min()\n",
    "\n",
    "sreads_df['column_crop'] = sreads_df['column_scaled'] - sreads_df['column_scaled'].min()\n",
    "sreads_df['row_crop'] = sreads_df['row_scaled'] - sreads_df['row_scaled'].min()\n",
    "\n",
    "bgreads_df['column_crop'] = bgreads_df['column_scaled'] - bgreads_df['column_scaled'].min()\n",
    "bgreads_df['row_crop'] = bgreads_df['row_scaled'] - bgreads_df['row_scaled'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4522fbef-884c-4860-8771-6f1067dffadd",
   "metadata": {},
   "outputs": [],
   "source": [
    "level_3_transfer_dict = dict(zip(sdata.obs.index.astype(int), sdata.obs['level_3']))\n",
    "sreads_df['level_3'] = sreads_df['orig_index'].map(level_3_transfer_dict)\n",
    "sreads_df.loc[sreads_df['level_3'].isna(), 'level_3'] = 'Unknown'\n",
    "sreads_df['level_3'] = sreads_df['level_3'].astype('category')\n",
    "\n",
    "subset_order = []\n",
    "subset_colors = []\n",
    "for i, current_cat in enumerate(sdata.uns['level_3_order']):\n",
    "   #  print(current_cat)\n",
    "    if current_cat in sreads_df['level_3'].values:\n",
    "        subset_order.append(current_cat)\n",
    "        subset_colors.append(sdata.uns['level_3_color_list'][i])\n",
    "        \n",
    "sreads_df['level_3'] = sreads_df['level_3'].cat.reorder_categories(subset_order)\n",
    "\n",
    "subset_pl = sns.color_palette(subset_colors)\n",
    "subset_cmap = ListedColormap(subset_pl.as_hex())\n",
    "sns.palplot(subset_pl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7502ec0a-8ef9-4889-828c-e1792f6f9bc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.collections import PatchCollection\n",
    "sample = 'RIBOmap'\n",
    "sample_key = f\"{sample}_morph\"\n",
    "nissl = sdata.uns[sample_key]['label_img_crop']\n",
    "hulls = sdata.uns[sample_key]['qhulls']\n",
    "\n",
    "sdata.uns[sample_key]['good_cells'] = sdata.obs.index.astype(int).values\n",
    "sdata.uns[sample_key]['colors'] = sdata.obs.loc[:, 'level_3'].cat.codes.values\n",
    "colors = sdata.uns[sample_key]['colors']\n",
    "good_cells = sdata.uns[sample_key]['good_cells']\n",
    "\n",
    "save_as_real_size = False\n",
    "figscale = 10\n",
    "if save_as_real_size:\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000, nissl.shape[1]/1000), dpi=100)\n",
    "else:\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000 * figscale, nissl.shape[1]/1000 * figscale), dpi=100)\n",
    "\n",
    "polys = []\n",
    "for h in hulls:\n",
    "    if h == []:\n",
    "        polys.append([])\n",
    "    else:\n",
    "        polys.append(su.hull_to_polygon(h))\n",
    "\n",
    "    \n",
    "if good_cells is not None:\n",
    "    others = [p for i, p in enumerate(polys) if i not in good_cells and p != []]\n",
    "    polys = [p for i, p in enumerate(polys) if i in good_cells]\n",
    "\n",
    "alpha = 1\n",
    "linewidth = .5\n",
    "p = PatchCollection(polys, alpha=alpha, cmap=level_3_cmap, edgecolor='k', linewidth=linewidth, zorder=3)\n",
    "\n",
    "other_cmap = sns.color_palette(['#ffffff']) \n",
    "other_cmap = ListedColormap(other_cmap)\n",
    "o = PatchCollection(others, alpha=1, cmap=other_cmap, edgecolor='k', linewidth=1, zorder=1)\n",
    "\n",
    "vmin = None\n",
    "vmax = None\n",
    "rescale_colors = False\n",
    "if vmin or vmax is not None:\n",
    "    p.set_array(colors)\n",
    "    p.set_clim(vmin=vmin, vmax=vmax)\n",
    "else:\n",
    "    if rescale_colors:\n",
    "        p.set_array(colors+1)\n",
    "        p.set_clim(vmin=0, vmax=max(colors+1))\n",
    "    else:\n",
    "        p.set_array(colors)\n",
    "        p.set_clim(vmin=0, vmax=len(level_3_cmap.colors))\n",
    "\n",
    "        o_colors = np.ones(len(others)).astype(int)\n",
    "        o.set_array(o_colors)\n",
    "        o.set_clim(vmin=0, vmax=max(o_colors))\n",
    "\n",
    "\n",
    "nissl = (nissl > 0).astype(np.int)\n",
    "plt.imshow(nissl.T, cmap=plt.get_cmap('gray_r'), alpha=0, zorder=1)\n",
    "\n",
    "# plt.gca().add_collection(p)\n",
    "plt.gca().add_collection(o)\n",
    "\n",
    "sns.scatterplot(x='row_crop', y='column_crop', hue='level_3', data=sreads_df, palette=subset_pl, s=3, legend=False, edgecolor=None, alpha=.5)\n",
    "\n",
    "plt.axis('off')\n",
    "plt.tight_layout(pad=0)\n",
    "\n",
    "current_fig_path = f\"{fig_path}/sct_RIBOmap_dots.tif\"\n",
    "plt.savefig(current_fig_path, bbox_inches='tight', pad_inches=0)\n",
    "plt.show()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5460fa5e-8561-4e85-90a1-ad8c06efacc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.patches import Polygon\n",
    "sample = 'RIBOmap'\n",
    "sample_key = f\"{sample}_morph\"\n",
    "nissl = sdata.uns[sample_key]['label_img_crop']\n",
    "hulls = sdata.uns[sample_key]['qhulls']\n",
    "\n",
    "sdata.uns[sample_key]['good_cells'] = sdata.obs.index.astype(int).values\n",
    "sdata.uns[sample_key]['colors'] = sdata.obs.loc[:, 'level_3'].cat.codes.values\n",
    "colors = sdata.uns[sample_key]['colors']\n",
    "good_cells = sdata.uns[sample_key]['good_cells']\n",
    "\n",
    "save_as_real_size = False\n",
    "figscale = 10\n",
    "if save_as_real_size:\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000, nissl.shape[1]/1000), dpi=100)\n",
    "else:\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000 * figscale, nissl.shape[1]/1000 * figscale), dpi=100)\n",
    "\n",
    "polys = []\n",
    "for h in hulls:\n",
    "    if h == []:\n",
    "        polys.append([])\n",
    "    else:\n",
    "        polys.append(su.hull_to_polygon(h))\n",
    "\n",
    "    \n",
    "if good_cells is not None:\n",
    "    others = [p for i, p in enumerate(polys) if i not in good_cells and p != []]\n",
    "    polys = [p for i, p in enumerate(polys) if i in good_cells]\n",
    "\n",
    "alpha = 1\n",
    "linewidth = .5\n",
    "p = PatchCollection(polys, alpha=alpha, cmap=level_3_cmap, edgecolor='k', linewidth=linewidth, zorder=3)\n",
    "\n",
    "other_cmap = sns.color_palette(['#ffffff']) \n",
    "other_cmap = ListedColormap(other_cmap)\n",
    "o = PatchCollection(others, alpha=1, cmap=other_cmap, edgecolor='k', facecolor=None, linewidth=1, zorder=1)\n",
    "\n",
    "vmin = None\n",
    "vmax = None\n",
    "rescale_colors = False\n",
    "if vmin or vmax is not None:\n",
    "    p.set_array(colors)\n",
    "    p.set_clim(vmin=vmin, vmax=vmax)\n",
    "else:\n",
    "    if rescale_colors:\n",
    "        p.set_array(colors+1)\n",
    "        p.set_clim(vmin=0, vmax=max(colors+1))\n",
    "    else:\n",
    "        p.set_array(colors)\n",
    "        p.set_clim(vmin=0, vmax=len(level_3_cmap.colors))\n",
    "\n",
    "        o_colors = np.ones(len(others)).astype(int)\n",
    "        o.set_array(o_colors)\n",
    "        o.set_clim(vmin=0, vmax=max(o_colors))\n",
    "\n",
    "\n",
    "nissl = (nissl > 0).astype(np.int)\n",
    "plt.imshow(nissl.T, cmap=plt.get_cmap('gray_r'), alpha=0, zorder=1)\n",
    "\n",
    "# plt.gca().add_collection(p)\n",
    "plt.gca().add_collection(o)\n",
    "\n",
    "sns.scatterplot(x='row_crop', y='column_crop', color='b', data=sreads_df, s=3, legend=False, edgecolor=None, alpha=.5)\n",
    "sns.scatterplot(x='row_crop', y='column_crop', color='r', data=bgreads_df, s=3, legend=False, edgecolor=None, alpha=.5)\n",
    "\n",
    "# plt.gca().add_collection(o)\n",
    "\n",
    "# for patch in others:\n",
    "#     plt.gca().add_patch(Polygon(patch.get_xy(), closed=True, ec='k', lw=1, fill=False))\n",
    "\n",
    "plt.axis('off')\n",
    "plt.tight_layout(pad=0)\n",
    "\n",
    "current_fig_path = f\"{fig_path}/sct_RIBOmap_dots_2.tif\"\n",
    "plt.savefig(current_fig_path, bbox_inches='tight', pad_inches=0)\n",
    "plt.show()   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81c558b0-7a9d-4532-b7b5-305cb7eb8579",
   "metadata": {
    "tags": []
   },
   "source": [
    "### gene expression plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8ddc955-f1a2-4b70-a1ef-0b3021fe980e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "n = 50\n",
    "save_as = True\n",
    "# plot_genes = adata.var.sort_values('total_counts', ascending=False).head(n).index.to_list()\n",
    "# plot_genes = plot_genes + ['Gad1', 'Cux2', 'Rorb', 'Pcp4', 'Gad2', 'Sst', 'Gfap']\n",
    "# plot_genes = ['Sst', 'Gad1', 'Plp1', 'Slc17a7', 'Pcp4', 'Gad2', 'Gfap']\n",
    "plot_genes = [\n",
    "'Gad1', 'Gad2', 'Slc32a1',\n",
    "'Slc17a7',\n",
    "'Vip', 'Sst', 'Chodl', 'Pvalb',\n",
    "'Cux2', 'Rorb', 'Pcp4', 'Nr4a2',\n",
    "'Aqp4', 'Aldoc', 'Gfap',\n",
    "'Mbp', 'Mobp', 'Plp1', 'Pdgfra',\n",
    "'Ctss', 'C1qa', 'Vtn', 'Bsg', 'Ptgds', 'Dcn',\n",
    "]\n",
    "\n",
    "sc.set_figure_params(format='tif', dpi=150)\n",
    "\n",
    "for gene in tqdm(plot_genes):\n",
    "    dot_size = .5\n",
    "    fig, ax = plt.subplots(figsize=(4, 5))\n",
    "    \n",
    "    # sns.scatterplot(x='column', y='row', data=cell_center_df, color='#fffffa', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    \n",
    "    current_expr = adata[:, gene].layers['norm'].flatten() # / adata[:, gene].layers['norm'].flatten().max()\n",
    "    sns.scatterplot(x='column', y='row', data=adata.obs, hue=current_expr, palette='Reds', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    # plt.title(gene)\n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    if save_as:\n",
    "        expr_fig_path = os.path.join(fig_path, f'{date}-expr-norm')\n",
    "        if not os.path.exists(expr_fig_path):\n",
    "            os.mkdir(expr_fig_path)\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{gene}.tif'))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5639989e-e89a-4ec5-a550-9865d59c4db6",
   "metadata": {
    "tags": []
   },
   "source": [
    "### gene expression plots w/ STARmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f806312d-1532-4572-913b-1f42937df0e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sdata = sc.read_h5ad(os.path.join(out_path, '2022-02-10-Hu-TissueSTARmap-mad-filtered.h5ad'))\n",
    "sdata.layers['raw'] = sdata.X\n",
    "sc.pp.normalize_total(sdata)\n",
    "sc.pp.log1p(sdata)\n",
    "sdata.layers['norm'] = sdata.X\n",
    "sdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83745b30-14b7-4ebf-ae11-bfe59904b6a6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "n = 50\n",
    "save_as = False\n",
    "expr_layer = 'norm'\n",
    "# plot_genes = adata.var.sort_values('total_counts', ascending=False).head(n).index.to_list()\n",
    "# plot_genes = plot_genes + ['Gad1', 'Cux2', 'Rorb', 'Pcp4', 'Gad2', 'Sst', 'Gfap']\n",
    "plot_genes = ['Sst', 'Gad1', 'Plp1', 'Slc17a7', 'Pcp4', 'Gad2', 'Gfap']\n",
    "sc.set_figure_params(format='tif', dpi=150)\n",
    "\n",
    "for gene in tqdm(plot_genes):\n",
    "    dot_size = .5\n",
    "    fig, axs = plt.subplots(figsize=(8, 5), ncols=2)\n",
    "    sns.scatterplot(x='column', y='row', data=cell_center_df, color='#fffffa', s=dot_size, legend=False, edgecolor=None, ax=axs[0])\n",
    "    \n",
    "    ribo_expr = adata[:, gene].layers[expr_layer].flatten() # / adata[:, gene].layers['norm'].flatten().max()\n",
    "    sns.scatterplot(x='column', y='row', data=adata.obs, hue=ribo_expr, palette='Reds', s=dot_size, legend='auto', edgecolor=None, ax=axs[0])\n",
    "    \n",
    "    star_expr = sdata[:, gene].layers[expr_layer].flatten() # / adata[:, gene].layers['norm'].flatten().max()\n",
    "    sns.scatterplot(x='column', y='row', data=sdata.obs, hue=star_expr, palette='Reds', s=dot_size, legend='auto', edgecolor=None, ax=axs[1])\n",
    "    axs[0].axis('off')\n",
    "    axs[1].axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    if save_as:\n",
    "        expr_fig_path = os.path.join(fig_path, 'expr_combined')\n",
    "        if not os.path.exists(expr_fig_path):\n",
    "            os.mkdir(expr_fig_path)\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{gene}.tif'))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "884dba91-b5ec-4246-aced-ca1a6b391f87",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "n = 50\n",
    "save_as = True\n",
    "expr_layer = 'norm'\n",
    "# plot_genes = adata.var.sort_values('total_counts', ascending=False).head(n).index.to_list()\n",
    "# plot_genes = plot_genes + ['Gad1', 'Cux2', 'Rorb', 'Pcp4', 'Gad2', 'Sst', 'Gfap']\n",
    "# plot_genes = ['Sst', 'Gad1', 'Plp1', 'Slc17a7', 'Pcp4', 'Gad2', 'Gfap']\n",
    "plot_genes = ['Ttc3','Hpca', 'Itpka', 'Pcp4', 'Arfgef1', 'Dcx', 'Map2', 'Nefl', 'Nptx1', 'Prkce', 'Rundc3a', 'Rap1gap', 'Scgn', 'Chat', 'Hcrt', 'Gpr151', 'Th', 'Slc18a3', 'Cbln3', 'Dpp6', 'Tbr1', 'Camkv',\n",
    "            'Adcyap1', 'Gal', 'Slc17a6', 'Sncg', 'Slc6a2', 'Aldoc', 'Gng2', 'Igfbp5', 'Lrpap1', 'Dact2', 'Arc', 'Emx2', 'Akap5', 'Cck', 'Necab1', 'Calb2', 'Slc30a3', 'Cartpt', 'Adora2a', 'Agrp', 'Avp',\n",
    "              'Cdh15', 'Cnp', 'Drd1', 'Drd2', 'Gng7', 'Gpr88', 'Homer3', 'Opt', 'Oxt', 'Pax5', 'Gs9', 'Slc1a6', 'Slc6a4', 'Slc6a5', 'Bcl11b', 'Calcr', 'Cnga3', 'Crabp1']\n",
    "plot_genes = [g for g in plot_genes if g in adata.var.index]\n",
    "sc.set_figure_params(format='tif', dpi=150)\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'{date}-expr-combined-{expr_layer}')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "            \n",
    "for gene in tqdm(plot_genes):\n",
    "    dot_size = .5\n",
    "    \n",
    "    # RIBOmap\n",
    "    fig, ax = plt.subplots(figsize=(4, 5))\n",
    "    sns.scatterplot(x='column', y='row', data=cell_center_df, color='#fffffa', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "\n",
    "    ribo_expr = adata[:, gene].layers[expr_layer].flatten() # / adata[:, gene].layers['norm'].flatten().max()\n",
    "    sns.scatterplot(x='column', y='row', data=adata.obs, hue=ribo_expr, palette='Reds', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    if save_as:\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{gene}_RIBOmap.tif'))\n",
    "    plt.show()\n",
    "    \n",
    "    # STARmap\n",
    "    fig, ax = plt.subplots(figsize=(4, 5))\n",
    "    sns.scatterplot(x='column', y='row', data=cell_center_df, color='#fffffa', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "\n",
    "    star_expr = sdata[:, gene].layers[expr_layer].flatten() # / adata[:, gene].layers['norm'].flatten().max()\n",
    "    sns.scatterplot(x='column', y='row', data=sdata.obs, hue=star_expr, palette='Reds', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    if save_as:\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{gene}_STARmap.tif'))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38dab98e-6be3-4ef6-85f7-b8b2cf96c018",
   "metadata": {
    "tags": []
   },
   "source": [
    "### gene expression plots (all reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94dbfed0-3a00-42c0-9249-be9aa88c47f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load reads \n",
    "good_reads = pd.read_csv(os.path.join(base_path, 'RIBOmap', 'remain_reads_polished.csv'), index_col=0)\n",
    "background_reads = pd.read_csv(os.path.join(base_path, 'RIBOmap', 'background_reads_polished.csv'), index_col=0)\n",
    "all_reads = pd.concat([good_reads, background_reads])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62e52fac-438a-4d05-860b-1ee6e20315fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load gene lists\n",
    "list_1 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_1')\n",
    "list_2 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_2')\n",
    "list_3 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_3')\n",
    "list_4 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_4')\n",
    "list_5 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_5')\n",
    "list_6 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_6')\n",
    "list_7 = pd.read_excel(os.path.join(base_path, 'gene-module', '2022-04-11-Hu-gene-list.xlsx'), sheet_name='list_7')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61767127-2009-4aa2-b1cd-2db4458c42fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "profiled_genes = all_reads['gene'].unique()\n",
    "print(len(profiled_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e142519-8e27-43e0-aef1-dfe78941d972",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_count = pd.DataFrame(all_reads['gene'].value_counts())\n",
    "reads_count.columns = ['count']\n",
    "top_500_genes = reads_count.iloc[:500, :]\n",
    "top_500_genes = top_500_genes.index.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aca78ae0-e8e8-4471-a2cf-c45ca41322cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_1\n",
    "save_as = True\n",
    "dot_size = 1\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'{date}-expr-reads', 'list-1')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "for i, current_gene in enumerate(tqdm(list_1['gene'])):\n",
    "    \n",
    "    if current_gene in top_500_genes:\n",
    "        \n",
    "        current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "        # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=all_reads, color='#ededed', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "        \n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(current_reads.loc[:,'spot_location_1'], current_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='r')\n",
    "\n",
    "        ax.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "\n",
    "        if save_as:\n",
    "            plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "            plt.close()\n",
    "        else:\n",
    "            plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8931b748-56f8-4a03-93c2-a68f90f8e690",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_2\n",
    "save_as = True\n",
    "dot_size = 1\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'{date}-expr-reads', 'list-2')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "for i, current_gene in enumerate(tqdm(list_2['gene'])):\n",
    "    \n",
    "    if current_gene in top_500_genes:\n",
    "        \n",
    "        current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "        # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=all_reads, color='#ededed', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "        \n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(current_reads.loc[:,'spot_location_1'], current_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='r')\n",
    "\n",
    "        ax.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "\n",
    "        if save_as:\n",
    "            plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "            plt.close()\n",
    "        else:\n",
    "            plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a87f30a-6623-4f3e-b1ab-316ff712619b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_3\n",
    "save_as = True\n",
    "dot_size = 1\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'{date}-expr-reads', 'list-3')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "for i, current_gene in enumerate(tqdm(list_3['gene'])):\n",
    "    \n",
    "    if current_gene in top_500_genes:\n",
    "        \n",
    "        current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "        # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=all_reads, color='#ededed', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "        \n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(current_reads.loc[:,'spot_location_1'], current_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='r')\n",
    "\n",
    "        ax.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "\n",
    "        if save_as:\n",
    "            plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "            plt.close()\n",
    "        else:\n",
    "            plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af429279-d066-4e4d-9772-42a198ee159b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_4\n",
    "save_as = True\n",
    "dot_size = 1\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'{date}-expr-reads', 'list-4')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "for i, current_gene in enumerate(tqdm(list_4['gene'])):\n",
    "    \n",
    "    if current_gene in top_500_genes:\n",
    "        \n",
    "        current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "        # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=all_reads, color='#ededed', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "        \n",
    "        sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "        # plt.scatter(current_reads.loc[:,'spot_location_1'], current_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='r')\n",
    "\n",
    "        ax.axis('off')\n",
    "        plt.tight_layout(pad=0)\n",
    "\n",
    "        if save_as:\n",
    "            plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "            plt.close()\n",
    "        else:\n",
    "            plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "737bedf9-b8b9-42d8-a5f0-cc077b8f2a49",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_gene = 'Atp1a1'\n",
    "current_soma_reads = good_reads.loc[good_reads['gene'] == current_gene, :]\n",
    "current_non_soma_reads = background_reads.loc[background_reads['gene'] == current_gene, :]\n",
    "\n",
    "dot_size = .3\n",
    "\n",
    "from matplotlib.collections import PatchCollection\n",
    "sample = 'RIBOmap'\n",
    "sample_key = f\"{sample}_morph\"\n",
    "nissl = adata.uns[sample_key]['label_img']\n",
    "hulls = adata.uns[sample_key]['qhulls']\n",
    "\n",
    "# colors = adata.uns[sample_key]['colors']\n",
    "colors = []\n",
    "good_cells = []\n",
    "\n",
    "save_as_real_size = True\n",
    "figscale = 10\n",
    "if save_as_real_size:\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000, nissl.shape[1]/1000), dpi=100)\n",
    "else:\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000 * figscale, nissl.shape[1]/1000 * figscale), dpi=100)\n",
    "\n",
    "polys = []\n",
    "for h in hulls:\n",
    "    if h == []:\n",
    "        polys.append([])\n",
    "    else:\n",
    "        polys.append(su.hull_to_polygon(h))\n",
    "\n",
    "    \n",
    "if good_cells is not None:\n",
    "    others = [p for i, p in enumerate(polys) if i not in good_cells and p != []]\n",
    "    polys = [p for i, p in enumerate(polys) if i in good_cells]\n",
    "\n",
    "alpha = 1\n",
    "linewidth = .1\n",
    "# p = PatchCollection(polys, alpha=alpha, cmap=level_3_cmap, edgecolor='k', linewidth=linewidth, zorder=3)\n",
    "\n",
    "other_cmap = sns.color_palette(['#ededed']) \n",
    "other_cmap = ListedColormap(other_cmap)\n",
    "o = PatchCollection(others, alpha=1, cmap=other_cmap, edgecolor='k', linewidth=0, zorder=1)\n",
    "\n",
    "vmin = None\n",
    "vmax = None\n",
    "rescale_colors = False\n",
    "# if vmin or vmax is not None:\n",
    "#     p.set_array(colors)\n",
    "#     p.set_clim(vmin=vmin, vmax=vmax)\n",
    "# else:\n",
    "#     if rescale_colors:\n",
    "#         p.set_array(colors+1)\n",
    "#         p.set_clim(vmin=0, vmax=max(colors+1))\n",
    "#     else:\n",
    "#         p.set_array(colors)\n",
    "#         p.set_clim(vmin=0, vmax=len(level_3_cmap.colors))\n",
    "\n",
    "o_colors = np.ones(len(others)).astype(int)\n",
    "o.set_array(o_colors)\n",
    "o.set_clim(vmin=0, vmax=max(o_colors))\n",
    "\n",
    "\n",
    "nissl = (nissl > 0).astype(np.int)\n",
    "plt.imshow(nissl.T, cmap=plt.get_cmap('gray_r'), alpha=0, zorder=1)\n",
    "\n",
    "# plt.gca().add_collection(p)\n",
    "plt.gca().add_collection(o)\n",
    "\n",
    "current_soma_reads['spot_location_1'] = current_soma_reads['spot_location_1'] * 0.3\n",
    "current_soma_reads['spot_location_2'] = current_soma_reads['spot_location_2'] * 0.3\n",
    "current_soma_reads['spot_location_1'] = current_soma_reads['spot_location_1'].astype(int)\n",
    "current_soma_reads['spot_location_2'] = current_soma_reads['spot_location_2'].astype(int)\n",
    "\n",
    "current_non_soma_reads['spot_location_1'] = current_non_soma_reads['spot_location_1'] * 0.3\n",
    "current_non_soma_reads['spot_location_2'] = current_non_soma_reads['spot_location_2'] * 0.3\n",
    "current_non_soma_reads['spot_location_1'] = current_non_soma_reads['spot_location_1'].astype(int)\n",
    "current_non_soma_reads['spot_location_2'] = current_non_soma_reads['spot_location_2'].astype(int)\n",
    "\n",
    "\n",
    "sns.scatterplot(x='spot_location_2', y='spot_location_1', data=current_non_soma_reads, color='r', s=dot_size, legend=False, edgecolor=None, markers='.')\n",
    "sns.scatterplot(x='spot_location_2', y='spot_location_1', data=current_soma_reads, color='b', s=dot_size, legend=False, edgecolor=None, markers='.')\n",
    "\n",
    "\n",
    "plt.axis('off')\n",
    "plt.tight_layout(pad=0)\n",
    "\n",
    "current_fig_path = f\"{fig_path}/test_{current_gene}.tif\"\n",
    "plt.savefig(current_fig_path, bbox_inches='tight', pad_inches=0)\n",
    "plt.show()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aebc8ddc-adc6-48bb-9f33-19b687e85a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_5\n",
    "save_as = True\n",
    "dot_size = 1\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'2022-04-12-expr-reads', 'list-5')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "for i, current_gene in enumerate(tqdm(list_5['gene'])):\n",
    "    \n",
    "    current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "    current_reads['spot_location_1'] = current_reads['spot_location_1'] * 0.3\n",
    "    current_reads['spot_location_2'] = current_reads['spot_location_2'] * 0.3\n",
    "    current_reads['spot_location_1'] = current_reads['spot_location_1'].astype(int)\n",
    "    current_reads['spot_location_2'] = current_reads['spot_location_2'].astype(int)\n",
    "    \n",
    "    sample = 'RIBOmap'\n",
    "    sample_key = f\"{sample}_morph\"\n",
    "    nissl = adata.uns[sample_key]['label_img']\n",
    "    hulls = adata.uns[sample_key]['qhulls']\n",
    "\n",
    "    colors = []\n",
    "    good_cells = []\n",
    "\n",
    "    plt.figure(figsize=(nissl.shape[0]/1000*3, nissl.shape[1]/1000*3), dpi=150)\n",
    "\n",
    "    polys = []\n",
    "    for h in hulls:\n",
    "        if h == []:\n",
    "            polys.append([])\n",
    "        else:\n",
    "            polys.append(su.hull_to_polygon(h))\n",
    "\n",
    "\n",
    "    if good_cells is not None:\n",
    "        others = [p for i, p in enumerate(polys) if i not in good_cells and p != []]\n",
    "        polys = [p for i, p in enumerate(polys) if i in good_cells]\n",
    "    \n",
    "\n",
    "    other_cmap = sns.color_palette(['#ededed']) \n",
    "    other_cmap = ListedColormap(other_cmap)\n",
    "    o = PatchCollection(others, alpha=1, cmap=other_cmap, edgecolor='k', linewidth=0, zorder=0)\n",
    "    o_colors = np.ones(len(others)).astype(int)\n",
    "    o.set_array(o_colors)\n",
    "    o.set_clim(vmin=0, vmax=max(o_colors))\n",
    "\n",
    "    nissl = (nissl > 0).astype(np.int)\n",
    "    plt.imshow(nissl.T, cmap=plt.get_cmap('gray_r'), alpha=0, zorder=-1)\n",
    "\n",
    "    plt.gca().add_collection(o)\n",
    "\n",
    "    sns.scatterplot(x='spot_location_2', y='spot_location_1', data=current_reads, color='r', s=dot_size, legend=False, edgecolor='r', zorder=3, markers='.')\n",
    "\n",
    "    plt.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "\n",
    "    current_fig_path = f\"{expr_fig_path}/{current_gene}.tif\"\n",
    "    plt.savefig(current_fig_path, bbox_inches='tight', pad_inches=0)\n",
    "    plt.close()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fbeb97a-d9ba-4991-aab4-00e204418399",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_5 with all dots\n",
    "save_as = True\n",
    "dot_size = 2\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'2022-04-12-expr-reads', 'list-5-dots-dark')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "genes = ['Shank1', 'Eef2', 'Plk2', 'Rtn4']\n",
    "# for i, current_gene in enumerate(tqdm(list_5['gene'])):\n",
    "for i, current_gene in enumerate(tqdm(genes)):\n",
    "    \n",
    "    current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', hue='soma', data=all_reads, s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax, zorder=0)\n",
    "    plt.plot(good_reads.loc[:,'spot_location_1'], good_reads.loc[:,'spot_location_2'], '.', ms=1, alpha=0.5, c='#ededed')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(current_reads.loc[:,'spot_location_1'], current_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='#2e0d52')\n",
    "\n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "\n",
    "    if save_as:\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc047b3a-dc9d-40e8-b42a-edbe21b9ef2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_5 with all dots two color\n",
    "save_as = True\n",
    "dot_size = 2\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'2022-04-12-expr-reads', 'list-5-dots-2')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "genes = ['Shank1', 'Eef2', 'Plk2', 'Rtn4']\n",
    "# for i, current_gene in enumerate(tqdm(list_5['gene'])):\n",
    "for i, current_gene in enumerate(tqdm(genes)):\n",
    "    \n",
    "    current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "    soma_reads = current_reads.loc[current_reads['cell_barcode'] != -1, :]\n",
    "    non_soma_reads = current_reads.loc[current_reads['cell_barcode'] == -1, :]\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', hue='soma', data=all_reads, s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax, zorder=0)\n",
    "    plt.plot(good_reads.loc[:,'spot_location_1'], good_reads.loc[:,'spot_location_2'], '.', ms=1, alpha=0.5, c='#ededed')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(soma_reads.loc[:,'spot_location_1'], soma_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(non_soma_reads.loc[:,'spot_location_1'], non_soma_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='r')\n",
    "\n",
    "    \n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "\n",
    "    if save_as:\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eae9e25a-e993-4356-b714-7997612b539a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_6 with all dots two color\n",
    "save_as = True\n",
    "dot_size = 4\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'2022-04-12-expr-reads', 'list-6-dots-4')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "genes = ['Eef2', 'Plk2', 'Rtn4', 'Shank1']\n",
    "# for i, current_gene in enumerate(tqdm(list_6['gene'])):\n",
    "for i, current_gene in enumerate(tqdm(genes)):\n",
    "    \n",
    "    current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "    soma_reads = current_reads.loc[current_reads['cell_barcode'] != -1, :]\n",
    "    non_soma_reads = current_reads.loc[current_reads['cell_barcode'] == -1, :]\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', hue='soma', data=all_reads, s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax, zorder=0)\n",
    "    plt.plot(good_reads.loc[:,'spot_location_1'], good_reads.loc[:,'spot_location_2'], '.', ms=1, alpha=0.5, c='#ededed')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(soma_reads.loc[:,'spot_location_1'], soma_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(non_soma_reads.loc[:,'spot_location_1'], non_soma_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='r')\n",
    "\n",
    "    \n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "\n",
    "    if save_as:\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30335676-c27e-48ae-82ce-34a8b3ba3aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot list_7 with all dots two color\n",
    "save_as = True\n",
    "dot_size = 4\n",
    "\n",
    "if save_as:\n",
    "    expr_fig_path = os.path.join(fig_path, f'2022-04-12-expr-reads', 'list-7-dots-4')\n",
    "    if not os.path.exists(expr_fig_path):\n",
    "        os.mkdir(expr_fig_path)\n",
    "\n",
    "# genes = ['Eef2', 'Plk2', 'Rtn4', 'Shank1']\n",
    "for i, current_gene in enumerate(tqdm(list_7['gene'][153:])):\n",
    "# for i, current_gene in enumerate(tqdm(genes)):\n",
    "    \n",
    "    current_reads = all_reads.loc[all_reads['gene'] == current_gene, :]\n",
    "    soma_reads = current_reads.loc[current_reads['cell_barcode'] != -1, :]\n",
    "    non_soma_reads = current_reads.loc[current_reads['cell_barcode'] == -1, :]\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=[40,50])\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', hue='soma', data=all_reads, s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "    # plt.scatter(all_reads.loc[:,'spot_location_1'], all_reads.loc[:,'spot_location_2'], s=1, alpha=0.2, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=good_reads, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax, zorder=0)\n",
    "    plt.plot(good_reads.loc[:,'spot_location_1'], good_reads.loc[:,'spot_location_2'], '.', ms=1, alpha=0.5, c='#ededed')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(soma_reads.loc[:,'spot_location_1'], soma_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='b')\n",
    "\n",
    "    # sns.scatterplot(x='spot_location_1', y='spot_location_2', data=current_reads, color='r', s=dot_size, legend=False, edgecolor=None, ax=ax, alpha=0.5, zorder=1)\n",
    "    plt.plot(non_soma_reads.loc[:,'spot_location_1'], non_soma_reads.loc[:,'spot_location_2'], '.', ms=dot_size, alpha=1, c='r')\n",
    "\n",
    "    \n",
    "    ax.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "\n",
    "    if save_as:\n",
    "        plt.savefig(os.path.join(expr_fig_path, f'{current_gene}_RIBOmap.tif'))\n",
    "        plt.close()\n",
    "    else:\n",
    "        plt.show()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0ab6a8a-bc20-478a-9fe3-d587a76babd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "67707b37-cffe-4bf3-8b69-f4bc48e724ec",
   "metadata": {},
   "source": [
    "### reads distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c4e3a2e-575c-421c-aec9-a01499840913",
   "metadata": {},
   "outputs": [],
   "source": [
    "good_reads_df = pd.DataFrame(good_reads['gene'].value_counts())\n",
    "good_reads_df.columns = ['soma reads']\n",
    "# good_reads_df['gene'] = good_reads_df.index\n",
    "\n",
    "background_reads_df = pd.DataFrame(background_reads['gene'].value_counts())\n",
    "background_reads_df.columns = ['non-soma reads']\n",
    "# background_reads_df['gene'] = background_reads_df.index\n",
    "\n",
    "all_reads_df = pd.DataFrame(all_reads['gene'].value_counts())\n",
    "all_reads_df.columns = ['all reads']\n",
    "# all_reads_df['gene'] = all_reads_df.index\n",
    "\n",
    "reads_df = pd.concat([good_reads_df, background_reads_df, all_reads_df], axis=1)\n",
    "reads_df['soma reads percentage'] = reads_df['soma reads'] / reads_df['all reads']\n",
    "reads_df['non-soma reads percentage'] = reads_df['non-soma reads'] / reads_df['all reads']\n",
    "reads_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7915913b-f5bf-45b6-811a-94186077ef7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ns_thres = reads_df['non-soma reads percentage'].quantile(0.90) # compare w/ list_1\n",
    "s_thres = reads_df['soma reads percentage'].quantile(0.90) # compare w/ list_4\n",
    "\n",
    "reads_df['top soma gene'] = reads_df['soma reads percentage'] > s_thres\n",
    "reads_df['top non-soma gene'] = reads_df['non-soma reads percentage'] > ns_thres\n",
    "reads_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b76e945-f7f1-4c6a-ae6b-3a3fc866d64d",
   "metadata": {},
   "outputs": [],
   "source": [
    "soma_gene_overlap = list_4.loc[list_4['gene'].isin(reads_df.loc[reads_df['top soma gene'], :].index.to_list()), :]\n",
    "non_soma_gene_overlap = list_1.loc[list_1['gene'].isin(reads_df.loc[reads_df['top non-soma gene'], :].index.to_list()), :]\n",
    "non_soma_gene_overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fccb07c5-0c0f-4b15-9979-6fe96ad7bd0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter(os.path.join(fig_path, f'{date}-reads-distribution.xlsx')) as writer:  \n",
    "    reads_df.to_excel(writer, sheet_name='reads-distribution')\n",
    "    soma_gene_overlap.to_excel(writer, sheet_name='soma-gene-overlap')\n",
    "    non_soma_gene_overlap.to_excel(writer, sheet_name='non-soma-gene-overlap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa3aef5-13dc-45b7-b8c7-5be71b4c0e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('white')\n",
    "fig, axs = plt.subplots(figsize=(10, 4), ncols=2)\n",
    "sns.histplot(reads_df['soma reads percentage'], ax=axs[0])\n",
    "axs[0].axvline(x=reads_df['soma reads percentage'].median(), color='r', linestyle='-')\n",
    "axs[0].set_title('soma reads percentage')\n",
    "\n",
    "sns.histplot(reads_df['non-soma reads percentage'], ax=axs[1])\n",
    "axs[1].axvline(x=reads_df['non-soma reads percentage'].median(), color='r', linestyle='-')\n",
    "axs[1].set_title('non-soma reads percentage')\n",
    "plt.savefig(os.path.join(fig_path, 'reads-distribution.pdf'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae585536-fe0d-4802-9259-59e33141217d",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_df['class'] = 0\n",
    "reads_df.loc[reads_df['top soma gene'] == True, 'class'] = 1\n",
    "reads_df.loc[reads_df['top non-soma gene'] == True, 'class'] = 2\n",
    "reads_df['class'] = reads_df['class'].astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54b02ac-8b7a-40ed-882b-780a97865d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "cpl_colors = ['#bfbfbf', '#1d43cf', '#cf1d1d']\n",
    "cpl = sns.color_palette(cpl_colors)\n",
    "cmap = ListedColormap(cpl.as_hex())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe93b55c-0e94-4486-b98f-0f3be544c8f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "fig, ax = plt.subplots(figsize=(6,5))\n",
    "\n",
    "sns.scatterplot(y='non-soma reads percentage', x='order', hue='class', data=reads_df, s=12, edgecolor=None, palette=cpl, legend=False)\n",
    "\n",
    "annotate_genes = ['Shank1', 'Eef2', 'Kif5a', 'Calm1', 'Gfap', 'Mbp', 'App', 'Rtn4', 'Mal']\n",
    "for gene in annotate_genes:\n",
    "    x = reads_df.loc[reads_df['index'] == gene, 'order'].values[0]\n",
    "    y = reads_df.loc[reads_df['index'] == gene, 'non-soma reads percentage'].values[0]\n",
    "    print(gene, x, y)\n",
    "    \n",
    "sns.scatterplot(y='non-soma reads percentage', x='order', hue='class', data=reads_df.loc[reads_df['index'].isin(annotate_genes), :], s=12, edgecolor='k', linewidth=1, palette=cpl, legend=False)\n",
    "    \n",
    "ax.annotate('Shank1', (2, 0.41132502274347815), xytext=(1000, 0.42), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Mbp', (4, 0.3922457104938857), xytext=(1000, 0.40), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Gfap', (20, 0.3576113516525914), xytext=(1000, 0.37), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Kif5a', (52, 0.3363509649236252), xytext=(1000, 0.35), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Eef2', (59, 0.33259025373758894), xytext=(1000, 0.33), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Calm1', (175, 0.3051969410410328), xytext=(1000, 0.30), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Rtn4', (4940, 0.14756159728122345), xytext=(4000, 0.14), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('App', (5073, 0.1385932516367299), xytext=(4000, 0.12), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "ax.annotate('Mal', (5378, 0.08249568623154666), xytext=(4000, 0.07), size=7,\n",
    "            bbox=dict(boxstyle=\"round\", alpha=0.1), \n",
    "            arrowprops = dict(arrowstyle='-', connectionstyle=\"arc3\", facecolor='black', edgecolor='black', lw=1)\n",
    "           )\n",
    "\n",
    "plt.savefig(os.path.join(fig_path, 'neuropil_reads.pdf'))\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "512a1ae9-9e35-42cd-9d72-2922c0f68ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "fig = px.scatter(reads_df, x=\"order\", y=\"non-soma reads percentage\", \n",
    "                 color='class', \n",
    "                 color_discrete_sequence=cpl_colors,\n",
    "                 category_orders={\"class\": [0, 1, 2]},\n",
    "                custom_data=['index', 'non-soma reads percentage', 'soma reads percentage'])\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=800,\n",
    "    height=800,)\n",
    "\n",
    "fig.update_traces(\n",
    "    hovertemplate=\"<br>\".join([\n",
    "        \"Gene: %{customdata[0]}\",\n",
    "        \"non-soma %: %{customdata[1]}\",\n",
    "        \"soma %: %{customdata[2]}\",\n",
    "    ])\n",
    ")\n",
    "\n",
    "fig.write_html(os.path.join(fig_path, 'neuropil_reads.html'))\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bec2bcf-b13f-40e0-908f-b1fc6c15b319",
   "metadata": {},
   "outputs": [],
   "source": [
    "# level 2 markers reads distribution \n",
    "marker_df = pd.read_csv(os.path.join(fig_path, 'level_2_markers.csv'), index_col=0)\n",
    "plot_marker_df = marker_df.groupby('group').head(10)\n",
    "plot_marker_df = plot_marker_df.loc[plot_marker_df['group'] != 'Unknown', :]\n",
    "plot_marker_df['non-soma reads percentage'] = 0\n",
    "print(plot_marker_df.shape)\n",
    "for gene in plot_marker_df['names']:\n",
    "    # print(gene)\n",
    "    plot_marker_df.loc[plot_marker_df['names'] == gene, 'non-soma reads percentage'] = reads_df.loc[reads_df['index'] == gene, 'non-soma reads percentage'].values[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeb32f24-a03b-488c-a42d-60e03580a1a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(7,3))\n",
    "ax = sns.violinplot(x=\"group\", y=\"non-soma reads percentage\", data=plot_marker_df, palette=\"tab10\",\n",
    "                     inner=\"box\")\n",
    "plt.xticks(rotation=45)\n",
    "# plt.savefig(os.path.join(fig_path, 'violin_ncounts_sample.pdf'))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4acb56-3420-4074-94d0-7b91c6a36509",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bff25ef0-1af4-4799-8e2f-a844ec615068",
   "metadata": {},
   "source": [
    "### Dendrogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "004c84ba-8b56-4537-a215-048e934cbddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "used_genes = adata.var.loc[adata.var['detected'] == True, :].index.to_list()\n",
    "sc.tl.dendrogram(adata, 'level_3', use_rep='X_pca', optimal_ordering=True, use_raw=True, n_pcs=30, var_names=used_genes)\n",
    "fig, ax = plt.subplots(figsize=(1,20))\n",
    "sc.set_figure_params(format='pdf', dpi=150)\n",
    "sc.pl.dendrogram(adata, 'level_3', orientation='left', ax=ax, save=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e10ecb44-ed8e-44a3-ac75-5fe9d341d0e6",
   "metadata": {},
   "source": [
    "### Sankey diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cfe94cc-1f7c-4b77-b3cd-c21c93d1c270",
   "metadata": {},
   "outputs": [],
   "source": [
    "def contingency(a, b, unique_a, unique_b):\n",
    "    \"\"\"Populate contingency matrix. Rows and columns are not normalized in any way.\n",
    "\n",
    "    Args:\n",
    "        a (np.array): labels\n",
    "        b (np.array): labels\n",
    "        unique_a (np.array): unique list of labels. Can have more entries than np.unique(a)\n",
    "        unique_b (np.array): unique list of labels. Can have more entries than np.unique(b)\n",
    "\n",
    "    Returns:\n",
    "        C (np.array): contingency matrix.\n",
    "    \"\"\"\n",
    "    # assert a.shape == b.shape\n",
    "    C = np.zeros((np.size(unique_a), np.size(unique_b)))\n",
    "    for i, la in enumerate(unique_a):\n",
    "        for j, lb in enumerate(unique_b):\n",
    "            C[i, j] = np.sum(np.logical_and(a == la, b == lb))\n",
    "            \n",
    "    df = pd.DataFrame(C)\n",
    "    df.index = unique_a\n",
    "    df.columns = unique_b\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd6f6d70-db95-4150-9a65-88c4399f0a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_nodes(df, key1, key2):\n",
    "    \n",
    "    df_river = {key1: [], key2: [], 'value': []}\n",
    "    for key2_type in df.columns.to_list():\n",
    "        for key1_type in df.index.to_list():\n",
    "            df_river[key1].append(key1_type)\n",
    "            df_river[key2].append(key2_type)\n",
    "            df_river['value'].append(df.loc[key1_type, key2_type])\n",
    "    df_river = pd.DataFrame(df_river)\n",
    "    df_river = df_river.loc[df_river['value'] != 0, :]\n",
    "\n",
    "    all_nodes = df_river[key1].unique().tolist() + df_river[key2].unique().tolist()\n",
    "    source_indices = [all_nodes.index(key1_type) for key1_type in df_river[key1]]\n",
    "    target_indices = [all_nodes.index(key2_type) for key2_type in df_river[key2]]\n",
    "    \n",
    "    return df_river, all_nodes, source_indices, target_indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "694eb187-6f2b-4efd-8d0f-b3bed59ebae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "level_12_df = contingency(adata.obs['level_1'].values, \n",
    "                   adata.obs['level_2'].values, \n",
    "                   adata.obs['level_1'].cat.categories.values, \n",
    "                   adata.obs['level_2'].cat.categories.values)\n",
    "level_12_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b711d893-dd68-4be9-ade3-4bfebb074aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_river, all_nodes, source_indices, target_indices = generate_nodes(level_12_df, 'level_1', 'level_2')\n",
    "\n",
    "level_12_colors = adata.uns['level_1_color_list'].tolist() + adata.uns['level_2_color_list'].tolist()\n",
    "node_colors_mappings = dict([(node, level_12_colors[i]) for i, node in enumerate(all_nodes)])\n",
    "node_colors = [node_colors_mappings[node] for node in all_nodes]\n",
    "edge_colors = [node_colors_mappings[node] for node in df_river['level_1']]\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "save_river = False\n",
    "fig = go.Figure(data=[go.Sankey(\n",
    "        node=dict(\n",
    "            pad=20,\n",
    "            thickness=20,\n",
    "            line=dict(color=\"black\", width=1.0),\n",
    "            label=all_nodes,\n",
    "            color=node_colors,\n",
    "        ),\n",
    "\n",
    "        link=dict(\n",
    "            source=source_indices,\n",
    "            target=target_indices,\n",
    "            value=df_river['value'],\n",
    "            color=edge_colors,\n",
    "        ))])\n",
    "\n",
    "fig.update_layout(title_text=\"level1 to level2\",\n",
    "                  height=600,\n",
    "                  font_size=10)\n",
    "if save_river:\n",
    "    fig.write_image(save_river)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9949dff2-fc8f-42ce-bd9f-d544396ea317",
   "metadata": {},
   "outputs": [],
   "source": [
    "level_23_df = contingency(adata.obs['level_2'].values, \n",
    "                   adata.obs['level_3'].values, \n",
    "                   adata.obs['level_2'].cat.categories.values, \n",
    "                   adata.obs['level_3'].cat.categories.values)\n",
    "level_23_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c930f79-3301-44e8-ab74-23bad8616341",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_river, all_nodes, source_indices, target_indices = generate_nodes(level_23_df, 'level_2', 'level_3')\n",
    "\n",
    "node_x = [0] * 8 + [1] * 57\n",
    "node_y = [i for i in range(8)] + [i*0.05 for i in range(57)]\n",
    "\n",
    "level_12_colors = adata.uns['level_2_color_list'].tolist() + adata.uns['level_3_color_list'].tolist()\n",
    "node_colors_mappings = dict([(node, level_12_colors[i]) for i, node in enumerate(all_nodes)])\n",
    "node_colors = [node_colors_mappings[node] for node in all_nodes]\n",
    "edge_colors = [node_colors_mappings[node] for node in df_river['level_2']]\n",
    "\n",
    "import plotly.graph_objects as go\n",
    "save_river = False\n",
    "fig = go.Figure(data=[go.Sankey(\n",
    "        node=dict(\n",
    "            pad=10,\n",
    "            thickness=20,\n",
    "            line=dict(color=\"black\", width=1.0),\n",
    "            label=all_nodes,\n",
    "            color=node_colors,\n",
    "            # x=node_x, \n",
    "            # y=node_y,\n",
    "        ),\n",
    "\n",
    "        link=dict(\n",
    "            source=source_indices,\n",
    "            target=target_indices,\n",
    "            value=df_river['value'],\n",
    "            color=edge_colors,\n",
    "        ))])\n",
    "\n",
    "fig.update_layout(\n",
    "                  height=3000,\n",
    "                  font_size=20)\n",
    "if save_river:\n",
    "    fig.write_image(save_river)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b5085e7-7a0b-43b6-a88b-edfcd4ebcad5",
   "metadata": {},
   "source": [
    "#### combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c31dd280-eb44-4452-a9fd-8afb06fd7ea2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get tables \n",
    "\n",
    "level_12_df = contingency(adata.obs['level_1'].values, \n",
    "                   adata.obs['level_2'].values, \n",
    "                   adata.obs['level_1'].cat.categories.values, \n",
    "                   adata.obs['level_2'].cat.categories.values)\n",
    "\n",
    "level_23_df = contingency(adata.obs['level_2'].values, \n",
    "                   adata.obs['level_3'].values, \n",
    "                   adata.obs['level_2'].cat.categories.values, \n",
    "                   adata.obs['level_3'].cat.categories.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1ae89ad-e5a1-4ce4-9b04-f92cc0ea115f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.uns['level_1_color_list_new'] = ['#eded58', '#356be8'] # ['#db5f57', '#57d3db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcac6bbb-500e-4f6e-add7-4fe88f277393",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get nodes\n",
    "\n",
    "df_river_12, all_nodes_12, source_indices_12, target_indices_12 = generate_nodes(level_12_df, 'level_1', 'level_2')\n",
    "df_river_23, all_nodes_23, source_indices_23, target_indices_23 = generate_nodes(level_23_df, 'level_2', 'level_3')\n",
    "\n",
    "all_nodes = adata.obs['level_1'].cat.categories.to_list() + adata.obs['level_2'].cat.categories.to_list() +  adata.obs['level_3'].cat.categories.to_list()\n",
    "source_indices = source_indices_12 + [i+2 for i in source_indices_23]\n",
    "target_indices = target_indices_12 + [i+2 for i in target_indices_23]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed83ef2-6c57-4182-aa76-8bcaedd46fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get colors and values \n",
    "all_colors = adata.uns['level_1_color_list_new'] + adata.uns['level_2_color_list'].tolist() + adata.uns['level_3_color_list'].tolist()\n",
    "node_colors_mappings = dict([(node, all_colors[i]) for i, node in enumerate(all_nodes)])\n",
    "node_colors = [node_colors_mappings[node] for node in all_nodes]\n",
    "edge_colors = [node_colors_mappings[node] for node in df_river_12['level_1']] + [node_colors_mappings[node] for node in df_river_23['level_2']]\n",
    "link_values = df_river_12['value'].to_list() + df_river_23['value'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf7c0fc9-382d-4c59-b601-4865ff00b3a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "save_river = False\n",
    "fig = go.Figure(data=[go.Sankey(\n",
    "        node=dict(\n",
    "            pad=10,\n",
    "            thickness=20,\n",
    "            line=dict(color=\"black\", width=1.0),\n",
    "            label=all_nodes,\n",
    "            color=node_colors,\n",
    "        ),\n",
    "\n",
    "        link=dict(\n",
    "            source=source_indices,\n",
    "            target=target_indices,\n",
    "            value=link_values,\n",
    "            # color=edge_colors,\n",
    "        ))])\n",
    "\n",
    "fig.update_layout(width=1500, \n",
    "                  height=2000,\n",
    "                  font_size=20)\n",
    "# if save_river:\n",
    "fig.write_image(os.path.join(fig_path, 'sankey-test.pdf'))\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5abac17f-28b9-476e-a1bf-d061a8f603bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c96ad27d-32b6-4a74-a4a4-8454ec0e125e",
   "metadata": {},
   "outputs": [],
   "source": [
    "coords_df4 = pd.read_csv('Z:/jiahao/Github/RIBOmap/segmentation-stitching/RIBOmap/results/tuned_coords.csv', index_col=0)\n",
    "coords_df4 = coords_df4.loc[coords_df4['tile'] != 0, :]\n",
    "coords_df4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7535128f-ddf0-464e-b73d-041f2aa8b3ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "tile_col = coords_df4['column'].values + 1000\n",
    "tile_row = coords_df4['row'].values + 1000\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(18,23))\n",
    "sns.scatterplot(x='column', y='row', data=cell_center_df, color='#dbdbdb', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "sns.scatterplot(x=tile_col, y=tile_row, color='black', s=dot_size, legend=False, edgecolor=None, ax=ax)\n",
    "for i in range(coords_df4.shape[0]):\n",
    "    plt.text(tile_col[i], tile_row[i], s=coords_df4['tile'].astype(str).to_list()[i], size=10, c='red')\n",
    "ax.set_aspect('equal')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5b28af9-d628-4b35-9edb-7b53effda5ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "coords_df4 = pd.read_csv('Z:/jiahao/Github/RIBOmap/segmentation-stitching/RIBOmap/results/tuned_coords.csv', index_col=0)\n",
    "coords_df4 = coords_df4.loc[coords_df4['tile'] != 0, :]\n",
    "\n",
    "col_min = 20000\n",
    "row_min = 45000\n",
    "col_max = 45000\n",
    "row_max = 69000\n",
    "\n",
    "corrds_logical = (coords_df4['column'].isin(range(col_min, col_max))) & (coords_df4['row'].isin(range(row_min, row_max)))\n",
    "scoords_df = coords_df4.loc[corrds_logical, :]\n",
    "scoords_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8422150-ecdf-4c8f-be63-f397c3c4bc94",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
